plugins/action/win_updates.py

```diff
@@ -756,6 +757,32 @@ def _run_sync(self, task_vars, module_options, reboot, reboot_timeout):  # type:
             self._download_results.update(update_result.download_results)
             self._install_results.update(update_result.install_results)
 
+            # Check that at least 1 update has not already been installed. This
+            # is to detect an update that may have been rolled back in the last
+            # reboot or whether WUA failed to report that the update failed.
+            same_updates = len(update_result.install_results) > 0
+            current_updates = set()
+            for update_id in update_result.install_results.keys():
+                if update_id in installed_updates:
+                    current_updates.add(update_id)
+                else:
+                    same_updates = False
+                    installed_updates.add(update_id)
```

Comment:
Since we've already got sets, isn't this better expressed in basic set operations, something like:

```suggestion
            current_updates = set(update_result.install_results.keys())
            duplicated_updates = current_updates.intersection(installed_updates)
            installed_updates.update(current_updates)
            
            if duplicated_updates:
            ...
```

Unless I'm misunderstanding the logic, the existing imperative code will also be "blind" to any duplicate that occurs that isn't the first non-duplicate, right? I don't think that's the desired behavior, so doing it all with sets also saves you from that bug.

Reply:
I've got the current logic so it's only if there are just duplicate updates. If it installed an update that hadn't been installed before it will still consider this round successful and try one more. That way if the current update somehow caused another to fail it will still try it by itself before considering it a lost cause.

Reply:
ah ok, so maybe like:

```suggestion
            current_updates = set(update_result.install_results.keys())
            new_updates = current_updates.difference(installed_updates)
            installed_updates.update(current_updates)
            
            if current_updates and not new_updates:
            ...
```

I'm probably being overly pedantic- it just feels more readable to me since the logic is purely set-oriented anyway.

Reply:
Yea makes sense, I'll swap it over to this new logic.